### PR TITLE
No reason to set transparency to typeclasses transparency when trying unification heuristics

### LINKED ISSUE #14374: The unification heuristics are applied in a context where "id" and "not" are frozen
#### Description of the problem

Unification heuristics use the [typeclasses flags](https://github.com/coq/coq/blob/34da65afbe8fec419498f113548a7a58638cf404/pretyping/pretyping.ml#L240) which make `id` and `not` [opaque](https://github.com/coq/coq/blob/34da65afbe8fec419498f113548a7a58638cf404/theories/Classes/Init.v#L22) (among others), leading to unexpected unification failures. (I don't have a reduced example though.)

See #14371 for a PR replacing the typeclasses transparency with the standard transparency.

#### Coq Version

Since 2011 (8.4)

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** inbetween bug fix and enhancement

Fixes / closes #14374

- [x] Added / updated test-suite
- [x] Entry added in the changelog

### CODE REVIEW

File: pretyping/pretyping.ml [hunk 1]
```diff
@@ -238,7 +238,7 @@ let apply_inference_hook (hook : inference_hook) env sigma frozen = match frozen
 
 let apply_heuristics env sigma fail_evar =
   (* Resolve eagerly, potentially making wrong choices *)
-  let flags = default_flags_of (Typeclasses.classes_transparent_state ()) in
+  let flags = default_flags_of TransparentState.full in
```
Comment (ppedrot) [inline pretyping/pretyping.ml:241]: Why not use the state from the environment instead? I know we have to keep backwards compatibility as much as possible, but I thought that tactics piercing Opaque directives were considered a bug.

Comment (herbelin) [reply pretyping/pretyping.ml]: OK, trying.

**Review by mattam82 — APPROVED** [review]
Nice fix!

**Comment by mattam82** [comment]
@coqbot merge now

**Pull request merged**